In [1]:
import pandas as pd
from matplotlib import pyplot as plt
import numpy as np
import math
import seaborn as sns
import sklearn
from sklearn import linear_model
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn import preprocessing
%matplotlib inline
sns.set_style('white')
pd.options.display.float_format = '{:.3f}'.format

from sklearn import ensemble
from sklearn import datasets
from sklearn.utils import shuffle
from sklearn.metrics import mean_squared_error

Suburb: Suburb

Address: Address

Rooms: Number of rooms

Price: Price in Australian dollars

Method: S - property sold; SP - property sold prior; PI - property passed in; PN - sold prior not disclosed; SN - sold not disclosed; NB - no bid; VB - vendor bid; W - withdrawn prior to auction; SA - sold after auction; SS - sold after auction price not disclosed. N/A - price or highest bid not available.

Type: br - bedroom(s); h - house,cottage,villa, semi,terrace; u - unit, duplex; t - townhouse; dev site - development site; o res - other residential.

SellerG: Real Estate Agent

Date: Date sold

Distance: Distance from CBD in Kilometres

Regionname: General Region (West, North West, North, North east ...etc)

Propertycount: Number of properties that exist in the suburb.

Bedroom2 : Scraped # of Bedrooms (from different source)

Bathroom: Number of Bathrooms

Car: Number of carspots

Landsize: Land Size in Metres

BuildingArea: Building Size in Metres

YearBuilt: Year the house was built

CouncilArea: Governing council for the area

Lattitude: Self explanitory

Longtitude: Self explanitory

In [2]:
housing_data = pd.read_csv('C:\\Users\\david\Desktop\\thinkful datasets\\Melbourne_housing_FULL.csv')
housing_data.head(5)

Suburb             Address  Rooms Type       Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h         nan     SS  Jellis   
1  Abbotsford        85 Turner St      2    h 1480000.000      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h 1035000.000      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u         nan     VB  Rounds   
4  Abbotsford        5 Charles St      3    h 1465000.000     SP  Biggin   

        Date  Distance  Postcode      ...       Bathroom   Car  Landsize  \
0  3/09/2016     2.500  3067.000      ...          1.000 1.000   126.000   
1  3/12/2016     2.500  3067.000      ...          1.000 1.000   202.000   
2  4/02/2016     2.500  3067.000      ...          1.000 0.000   156.000   
3  4/02/2016     2.500  3067.000      ...          2.000 1.000     0.000   
4  4/03/2017     2.500  3067.000      ...          2.000 0.000   134.000   

   BuildingArea  YearBuilt         CouncilArea Lattitude  Longtitude  \
0           nan        nan  Yarra City Council   -37.801     144.996   
1           nan        nan  Yarra City Council   -37.800     144.998   
2        79.000   1900.000  Yarra City Council   -37.808     144.993   
3           nan        nan  Yarra City Council   -37.811     145.012   
4       150.000   1900.000  Yarra City Council   -37.809     144.994   

              Regionname Propertycount  
0  Northern Metropolitan      4019.000  
1  Northern Metropolitan      4019.000  
2  Northern Metropolitan      4019.000  
3  Northern Metropolitan      4019.000  
4  Northern Metropolitan      4019.000  

[5 rows x 21 columns]

In [3]:
housing_data.describe()

Rooms        Price  Distance  Postcode  Bedroom2  Bathroom  \
count 34857.000    27247.000 34856.000 34856.000 26640.000 26631.000   
mean      3.031  1050173.345    11.185  3116.063     3.085     1.625   
std       0.970   641467.130     6.789   109.024     0.981     0.724   
min       1.000    85000.000     0.000  3000.000     0.000     0.000   
25%       2.000   635000.000     6.400  3051.000     2.000     1.000   
50%       3.000   870000.000    10.300  3103.000     3.000     2.000   
75%       4.000  1295000.000    14.000  3156.000     4.000     2.000   
max      16.000 11200000.000    48.100  3978.000    30.000    12.000   

            Car   Landsize  BuildingArea  YearBuilt  Lattitude  Longtitude  \
count 26129.000  23047.000     13742.000  15551.000  26881.000   26881.000   
mean      1.729    593.599       160.256   1965.290    -37.811     145.002   
std       1.011   3398.842       401.267     37.328      0.090       0.120   
min       0.000      0.000         0.000   1196.000    -38.190     144.424   
25%       1.000    224.000       102.000   1940.000    -37.863     144.934   
50%       2.000    521.000       136.000   1970.000    -37.808     145.008   
75%       2.000    670.000       188.000   2000.000    -37.754     145.072   
max      26.000 433014.000     44515.000   2106.000    -37.390     145.526   

       Propertycount  
count      34854.000  
mean        7572.888  
std         4428.090  
min           83.000  
25%         4385.000  
50%         6763.000  
75%        10412.000  
max        21650.000

In [4]:
# Counting nulls in the dataset 
null_count = housing_data.isnull().sum()
null_count[null_count>0]

Price             7610
Distance             1
Postcode             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
YearBuilt        19306
CouncilArea          3
Lattitude         7976
Longtitude        7976
Regionname           3
Propertycount        3
dtype: int64

In [5]:
housing_data = housing_data[housing_data["Price"].notnull()]

In [6]:
housing_data.head(5)

Suburb           Address  Rooms Type       Price Method SellerG  \
1  Abbotsford      85 Turner St      2    h 1480000.000      S  Biggin   
2  Abbotsford   25 Bloomburg St      2    h 1035000.000      S  Biggin   
4  Abbotsford      5 Charles St      3    h 1465000.000     SP  Biggin   
5  Abbotsford  40 Federation La      3    h  850000.000     PI  Biggin   
6  Abbotsford       55a Park St      4    h 1600000.000     VB  Nelson   

        Date  Distance  Postcode      ...       Bathroom   Car  Landsize  \
1  3/12/2016     2.500  3067.000      ...          1.000 1.000   202.000   
2  4/02/2016     2.500  3067.000      ...          1.000 0.000   156.000   
4  4/03/2017     2.500  3067.000      ...          2.000 0.000   134.000   
5  4/03/2017     2.500  3067.000      ...          2.000 1.000    94.000   
6  4/06/2016     2.500  3067.000      ...          1.000 2.000   120.000   

   BuildingArea  YearBuilt         CouncilArea Lattitude  Longtitude  \
1           nan        nan  Yarra City Council   -37.800     144.998   
2        79.000   1900.000  Yarra City Council   -37.808     144.993   
4       150.000   1900.000  Yarra City Council   -37.809     144.994   
5           nan        nan  Yarra City Council   -37.797     144.997   
6       142.000   2014.000  Yarra City Council   -37.807     144.994   

              Regionname Propertycount  
1  Northern Metropolitan      4019.000  
2  Northern Metropolitan      4019.000  
4  Northern Metropolitan      4019.000  
5  Northern Metropolitan      4019.000  
6  Northern Metropolitan      4019.000  

[5 rows x 21 columns]

In [7]:
housing_data.describe()

Rooms        Price  Distance  Postcode  Bedroom2  Bathroom  \
count 27247.000    27247.000 27246.000 27246.000 20806.000 20800.000   
mean      2.992  1050173.345    11.280  3113.796     3.046     1.592   
std       0.955   641467.130     6.787   111.140     0.955     0.701   
min       1.000    85000.000     0.000  3000.000     0.000     0.000   
25%       2.000   635000.000     6.400  3046.000     2.000     1.000   
50%       3.000   870000.000    10.500  3088.000     3.000     1.000   
75%       4.000  1295000.000    14.000  3153.000     4.000     2.000   
max      16.000 11200000.000    48.100  3978.000    20.000     9.000   

            Car   Landsize  BuildingArea  YearBuilt  Lattitude  Longtitude  \
count 20423.000  17982.000     10656.000  12084.000  20993.000   20993.000   
mean      1.715    593.489       156.835   1966.609    -37.807     144.997   
std       0.994   3757.266       449.223     36.762      0.092       0.121   
min       0.000      0.000         0.000   1196.000    -38.190     144.424   
25%       1.000    219.000       100.000   1950.000    -37.861     144.925   
50%       2.000    512.000       133.000   1970.000    -37.800     145.003   
75%       2.000    664.000       182.000   2000.000    -37.749     145.069   
max      18.000 433014.000     44515.000   2019.000    -37.398     145.526   

       Propertycount  
count      27244.000  
mean        7566.781  
std         4492.382  
min           83.000  
25%         4294.000  
50%         6567.000  
75%        10412.000  
max        21650.000

In [8]:
housing_data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [79]:
X= housing_data[['Rooms', 'Type',
        'Distance', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea']]

y = housing_data['Price']

In [80]:
X = X.fillna(X.median())

In [81]:
X = pd.get_dummies(X)

In [82]:
#X.loc[:,"Type_u"]
X.loc[lambda df: X['Type_u'] > 0, :]

Rooms  Distance  Bathroom   Car  Landsize  BuildingArea  Type_h  \
17         1     2.500     1.000 1.000     0.000       133.000       0   
19         2     2.500     1.000 2.000   512.000       133.000       0   
21         2     2.500     1.000 2.000   512.000       133.000       0   
22         1     2.500     1.000 2.000   512.000       133.000       0   
23         2     2.500     2.000 1.000     0.000       133.000       0   
25         2     2.500     2.000 1.000     0.000        94.000       0   
27         1     2.500     1.000 2.000   512.000       133.000       0   
29         1     2.500     1.000 1.000     0.000       133.000       0   
41         3     2.500     2.000 2.000  4290.000        27.000       0   
42         2     2.500     1.000 1.000     0.000        60.000       0   
45         1     2.500     1.000 2.000   512.000       133.000       0   
46         1     2.500     1.000 2.000   512.000       133.000       0   
48         1     2.500     1.000 1.000     0.000       133.000       0   
50         1     2.500     1.000 1.000     0.000        50.000       0   
57         1     2.500     1.000 1.000     0.000        50.000       0   
62         2     2.500     1.000 2.000   512.000       133.000       0   
73         2    13.500     1.000 1.000   168.000        86.000       0   
83         2    13.500     1.000 1.000   141.000        73.000       0   
85         2    13.500     1.000 1.000   120.000        94.000       0   
87         3    13.500     1.000 2.000   512.000       133.000       0   
90         3    13.500     1.000 1.000   265.000       130.000       0   
91         3    13.500     1.000 2.000   512.000       133.000       0   
94         2    13.500     1.000 1.000   192.000       133.000       0   
99         2    13.500     1.000 2.000   512.000       133.000       0   
101        2    13.500     1.000 1.000   164.000       133.000       0   
115        3    13.500     1.000 2.000   512.000       133.000       0   
117        2    13.500     1.000 1.000   141.000        76.000       0   
123        2    13.500     1.000 2.000   512.000       133.000       0   
127        3    13.500     1.000 2.000   512.000       133.000       0   
130        3    13.500     1.000 2.000   512.000       133.000       0   
...      ...       ...       ...   ...       ...           ...     ...   
34669      2     2.400     1.000 2.000   512.000       133.000       0   
34674      2    21.300     1.000 2.000   512.000       133.000       0   
34677      1     7.200     1.000 2.000   512.000       133.000       0   
34692      2    15.200     1.000 2.000   512.000       133.000       0   
34694      2    35.400     1.000 2.000   512.000       133.000       0   
34700      2     1.900     1.000 2.000   512.000       133.000       0   
34706      3     1.900     1.000 2.000   512.000       133.000       0   
34709      2     2.700     1.000 2.000   512.000       133.000       0   
34710      2     2.700     1.000 2.000   512.000       133.000       0   
34712      3     2.700     1.000 2.000   512.000       133.000       0   
34715      2     2.700     1.000 2.000   512.000       133.000       0   
34716      2     2.700     1.000 2.000   512.000       133.000       0   
34717      3     0.700     1.000 2.000   512.000       133.000       0   
34719      3     0.700     1.000 2.000   512.000       133.000       0   
34732      2     5.000     1.000 2.000   512.000       133.000       0   
34735      1     5.000     1.000 2.000   512.000       133.000       0   
34736      1     5.000     1.000 2.000   512.000       133.000       0   
34738      2     5.000     1.000 2.000   512.000       133.000       0   
34740      3     8.200     1.000 2.000   512.000       133.000       0   
34768      2    10.200     1.000 2.000   512.000       133.000       0   
34772      3    18.000     1.000 2.000   512.000       133.000       0   
34778      3    12.400     1.000 2.000   512.000       133.000       0   
34794      

In [84]:

# Instantiate and fit our model.# Instan 
regr = linear_model.LinearRegression()
#Y = data['Sales'].values.reshape(-1, 1)
#X = data[['TV','Radio','Newspaper']]
regr.fit(X, y)

# Inspect the results.
print('\nCoefficients: \n', regr.coef_)
print('\nIntercept: \n', regr.intercept_)
print('\nR-squared:')
print(regr.score(X, y))


Coefficients: 
 [ 2.25689883e+05 -3.75964889e+04  1.87400913e+05  3.97430016e+04
  4.46937987e+00  5.36403231e+01  1.82870881e+05 -3.29108729e+04
 -1.49960008e+05]

Intercept: 
 357730.92537712306

R-squared:
0.4085056495559721


In [85]:
coef_dict = {}
for coef, feat in zip(regr.coef_,X):
    coef_dict[feat] = coef

In [86]:
coef_dict

{'Rooms': 225689.88284121585,
 'Distance': -37596.48893590972,
 'Bathroom': 187400.91303890842,
 'Car': 39743.00155796949,
 'Landsize': 4.469379865317023,
 'BuildingArea': 53.640323147956224,
 'Type_h': 182870.88084475117,
 'Type_t': -32910.872867684215,
 'Type_u': -149960.00797706642}

In [21]:
"""Using this Kaggle data create a model to predict a house's value. We want to be able to understand what creates value in a house, as though we were a real estate developer."""

"Using this Kaggle data create a model to predict a house's value. We want to be able to understand what creates value in a house, as though we were a real estate developer."

- Base value of a home is \$ 357730.92
- Each room adds a value of \$ 225689.88
- Distance from CBD affects housing price by \$ 33,800
- Bathroom adds a value of \$ 187400.91
- Each car spot adds \$ 39743.00
- Each square meter adds \$ 4.46 of value
- Building area adds \$ 53.64 in value
- If the property is a house the value goes up \$ 182870 of the base value
- If the property is a townhouse the value goes down \$ 2940 of the base value
- If the property is an apartment the value goes down \$ 49960 of the base value